internal/impl/nats/processor_kv.go

```diff
@@ -286,18 +311,35 @@ func (p *kvProcessor) Process(c context.Context, msg *service.Message) (service.
 		}
 		return batch, nil
 
-	case kvKeys:
-		keys, err := kv.Keys()
+	case kvpOperationKeys:
+		// `kv.ListKeys()` does not allow users to specify a key filter, so we call `kv.Watch()` directly.
+		watcher, err := kv.Watch(p.keysFilter, []nats.WatchOpt{nats.IgnoreDeletes(), nats.MetaOnly()}...)
 		if err != nil {
 			return nil, err
 		}
-		batch := service.MessageBatch{}
-		for _, key := range keys {
-			m := service.NewMessage([]byte(key))
-			m.MetaSetMut(metaKVBucket, p.bucket)
-			batch = append(batch, m)
+		defer func() {
+			_ = watcher.Stop()
+		}()
+
+		var keys []any
+	loop:
+		for {
+			select {
+			case entry := <-watcher.Updates():
+				if entry == nil {
+					break loop
+				}
+				keys = append(keys, entry.Key())
+			case <-ctx.Done():
+				return nil, fmt.Errorf("watcher update loop exited prematurely: %s", ctx.Err())
+			}
 		}
```

Comment:
I'm a bit concerned about this approach, because it seems to block indefinitely in some cases. For example, try setting `keys_filter: '"001.>"'`. Not sure if this is by design, or a bug, but I think we might need to consider adding a `timeout` field that lets us error out if a timer expires while we wait for updates. WDYT?

Reply:
Interesting you are getting blocking on that value, I wonder if something is breaking when you send NATS a subject it doesn't like.

All that said, it's entirely possible for the watcher to never end, depending on how fast keys are being inserted. More of an edge case, but one concession we could make is to either have a configurable timeout, or stop the watcher when a specific time is reached, or a combination of the two.

I don't particularly like the idea that a simple timeout would result in receiving a partial set of keys. so I agree an error would be best in the simple timeout case.

That said, I kinda do like the idea of just returning the keys up to time that we requested it, would result in potentially more stale data for more timeliness on the retrieval side.

This is definitely more of an edge case where we have a really hot kv at the moment the watcher is created. I'm cool with any of the solutions.

Reply:
I think I'll go with the more conservative route and just return an error if the timeout lapses. People should take that as a sign that they either need to bump up the timeout or use a wildcard which selects fewer keys. We could enable partial results later on with a separate flag if anyone needs that functionality.

I'll slap together some quick code which shows how `'"001.>"'` blocks and I'll open an issue here https://github.com/nats-io/nats.go.

Reply:
Issue opened here: https://github.com/nats-io/nats.go/issues/1608

Reply:
And I also added a `timeout` field in 83c662e.

Reply:
I think that's a good solution for most cases!